src/devices/sound/rf5c400.cpp

```diff
@@ -344,6 +367,32 @@ uint16_t rf5c400_device::rf5c400_r(offs_t offset, uint16_t mem_mask)
 				return 0;
 			}
 
+			case 0x09:      // position read?
+			{
+				// The game will always call rf5c400_w 0x08 with a channel number and some other value before reading this register.
+				// The call to rf5c400_w 0x08 contains additional information, potentially what information it's expecting to be returned here.
+				// This implementation assumes all commands want the same information as command 6.
+				m_stream->update();
+
+				rf5c400_channel* channel = &m_channels[m_req_channel];
+				if (channel->env_phase == PHASE_NONE)
+				{
+					return 0;
+				}
+
+				// pop'n music's SPU program expects to read this register 6 times with the same value between every read before it will send the next DMA request.
+				//
+				// This register is polled while a streaming BGM is being played.
+				// For pop'n music specifically, the game starts off by reading 0x200000 into 0x00780000 - 0x00880000.
+				// When 2xxx is found (pos - start = 0x00080000), it will trigger the next DMA of 0x100000 overwriting 0x00780000 - 0x00800000, and continues polling the register until it reads 1xxx next.
+				// When 1xxx is found (pos - start = 0x00040000), it will trigger the next DMA of 0x100000 overwriting 0x00800000 - 0x00880000, and continues polling the register until it reads 2xxx next.
+				// ... repeat until song is finished, alternating between 2xxx and 1xxx ...
+				// This ends up so that it'll always be buffering new sample data into the sections of memory that aren't being used.
+				auto start = ((channel->startH & 0xFF00) << 8) | channel->startL;
+				auto ch_offset = (channel->pos >> 16) - start;
+				return ch_offset >> 6;
```

Comment:
I get a bit nervous with code like this – `channel->startH` is a `uint16_t`, but this relies on C’s implicit promotion to `unsigned int` by the `&` operator, and the fact that we only build for data type models that have at least 32-bit `int1`.  Could you change it to:
```
auto start = (uint32_t(channel->startH & 0xFF00) << 8) | channel->startL;
```
jut to make the intended promotion explicit.

Reply:
Understandable. This particular code is copy/pasted from another part of the code so I'll make the change for all such usages for consistency.

Reply:
You don’t actually need the extra parentheses – this *is* C++ :D

Reply:
That makes me want to read `uint32_t` like a function which my brain just won't allow. :P

Reply:
It’s a call to a `uint32_t` constructor.  More seriously though, it's less cluttered, easier to read and works better with precedence rules because it’s not right-associative.

Consider:
```
((foo)(x->y)).z();
foo(x->y).z();
```

In the first case, you need an extra set of parentheses because the C-style cast is right-associative and will have lower precedence than the `.` member access.

Reply:
That actually makes more sense when shown that code example. I'll keep that in mind for next time.